### ISSUE TITLE
[ECS] [Deployment]: should update Cloudformation events and template when trigger deployment circuit breaker and rollback

### ISSUE THREAD

**Issue by forward2you** [aws/containers-roadmap#1205]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
What do you want us to build?

ECS Fargate background task

**Which service(s) is this request for?**
ECS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

We use Cloudformation to deploy Fargate background task and adopt the deployment circuit breaker to avoid Cloudformation stuck in **UPDATE_IN_PROGRESS** if the task runs with a problematic container.
When triggered circuit breaker and rollback, the rollback status could only show in ECS service events, and by task definition version, we could verify that task has rollback.
However after rollback complete, the status in the Cloudformation would enter **UPDATE_COMPLETE**, and the Cloudformation template is still the problematic configuration, which would cause inconsistent between current ECS Service and Cloudformation stack.

ECS Service Events:
```
2020-12-24 15:10:20 +0800 service has reached a steady state.
2020-12-24 15:10:20 +0800 service (deployment ecs-svc/xxx) deployment completed.
2020-12-24 15:09:19 +0800 service stopped 1 pending tasks.
2020-12-24 15:09:09 +0800 service rolling back to deployment ecs-svc/xxx.
2020-12-24 15:09:09 +0800 service (deployment ecs-svc/xxx) deployment failed: tasks failed to start.
```

Cloudformation Events:
```
| 2020-12-24 15:10:15 UTC+0800 | demo | UPDATE_COMPLETE | -
| 2020-12-24 15:10:15 UTC+0800 | EcsTaskDefinition | DELETE_COMPLETE | -
| 2020-12-24 15:10:13 UTC+0800 | EcsTaskDefinition | DELETE_IN_PROGRESS | -
| 2020-12-24 15:10:12 UTC+0800 | demo | UPDATE_COMPLETE_CLEANUP_IN_PROGRESS | -
| 2020-12-24 15:10:08 UTC+0800 | EcsService | UPDATE_COMPLETE | -
<<<<<<<<<<Triggered circuit breaker and rollback>>>>>>>>>>>>>>>
| 2020-12-24 14:48:00 UTC+0800 | EcsService | UPDATE_IN_PROGRESS | -
| 2020-12-24 14:47:58 UTC+0800 | EcsTaskDefinition | UPDATE_COMPLETE | -
| 2020-12-24 14:47:58 UTC+0800 | EcsTaskDefinition | UPDATE_IN_PROGRESS | Resource creation Initiated
| 2020-12-24 14:47:55 UTC+0800 | EcsTaskDefinition | UPDATE_IN_PROGRESS | Requested update requires the creation of a new physical resource; hence creating one.
| 2020-12-24 14:47:48 UTC+0800 | demo | UPDATE_IN_PROGRESS
```

**What's our expectation on the status change in Cloudformation?**

I think after rollback, the Cloudformation status should be: 
  UPDATE_IN_PROGRESS-->UPDATE_FAILED-->UPDATE_ROLLBACK_IN_PROGRESS-->UPDATE_ROLLBACK_COMPLETE
And revert the cloudformation template to the latest configuration. By that way, user would know the ECS task update failed.

**Comment by fbau3r** [comment]
I tried the following configuration with my `AWS::ECS::Service` (though I'm running the service on ECS EC2 cluster, not Fargate ECS):
```yaml
DeploymentCircuitBreaker:
  Enable: True
  Rollback: False
```

I thought that this could be the way to configure that the ECS Orchestrator shall NOT rollback automatically and then CloudFormation picks up the triggered CircuitBreaker event and rolls back the current Stack Update.

Actual behavior seems unexpected: Stack Update is running, CircuitBreaker triggers (but does not roll back or cause an error). ECS Orchestrator does not try to deploy the new Task anymore and Stack Update waits for the new Task forever (kind of like before the CircuitBreaker but without ECS spawning failing Tasks).

**Comment by ArnoldKell** [comment]
We tried out deployment circuit breakers and ended up running into the same issue. As @fbau3r mentioned, setting the option to false doesn't fix it either (makes it worse, really) so currently it doesn't solve the pain point that we thought it was designed to solve. I'd also warn anyone that wants to try this, that you can also get stuck in the UPDATE_ROLLBACK_IN_PROGRESS status when experimenting with the circuit breakers.

**Comment by MikeKroell** [comment]
We just ran into this issue. It makes things worse. Back to timeout on the cloud formation.

**Comment by mwarkentin** [comment]
We're hitting this too, expected the ECS service deployment failure to show up as a failure in Cloudformation as well.

**Comment by mwarkentin** [comment]
From the docs it doesn't seem to mention Cloudformation as a supported configuration option - maybe there's a chance this is a partially complete feature release (and not just a documentation gap which seems more likely):

> The deployment circuit breaker can only be configured for a service using the Amazon ECS API, AWS CLI, or SDK.

**Comment by pavneeta** [comment]
Hi Everyone, 
Thank you for your valuable feedback. This is a bug in the CFN integration of the Deployment Circuit Breaker feature that we are actively tracking . We are currently working on a resolution.

**Comment by pavneeta** [comment]
Hi Everyone, 

We have rolled out a fix for the CloudFormation integration bugs for the Deployment Circuit Breaker feature across all AWS commercial regions. 

A deployment failure triggered due to the deployment circuit breaker with no rollback enabled now fails the CFN stack update/create with the following error `Error occurred during operation 'ECS Deployment Circuit Breaker was triggered.`

In case Rollback is enabled, ECS will rollback the deployment to previous service version and signal to CloudFormation once rollback is complete. This  will then be picked up by CFN to rollback the stack and the stack will end up with the status UPDATE_ROLLBACK_COMPLETE.

Please leave any additional feedback regarding this issue here.

Thanks.

**Comment by mwarkentin** [comment]
Initial tests of this functionality seem to be working great now, thanks!

**Comment by jakajancar** [comment]
@pavneeta I'm not sure from your description whether I should enable rollback on the circuit breaker, or rely on CloudFormation's rollback and leave ECS's disabled.

**Comment by adrian-skybaker** [comment]
> A deployment failure triggered due to the deployment circuit breaker with no rollback enabled now fails the CFN stack update/create with the following error `Error occurred during operation 'ECS Deployment Circuit Breaker was triggered.`
> 
> In case Rollback is enabled, ECS will rollback the deployment to previous service version and signal to CloudFormation once rollback is complete. This will then be picked up by CFN to rollback the stack and the stack will end up with the status UPDATE_ROLLBACK_COMPLETE.

Is this documented somewhere? If not, could it be added to the documentation?

**Comment by cbackas** [comment]
> @pavneeta I'm not sure from your description whether I should enable rollback on the circuit breaker, or rely on CloudFormation's rollback and leave ECS's disabled.

Did you (or anyone else) determine whether its best to have ECS perform the rollback or rely on CF?

**Comment by adrian-skybaker** [comment]
> @jakajancar Did you (or anyone else) determine whether its best to have ECS perform the rollback or rely on CF?

My experience is that setting `Rollback: true` definitely did *not* result in the behaviour we wanted. The deployment is roll back "silently" and the cloudformation deployment is marked as successful. So you end up with a green deploy pipeline, not realising the new version failed to deploy. I didn't try it, but I also suspect attempting to "redeploy" at this stage would result in an empty changeset, since as far as cloudformation is concerned, there is nothing to do.
